### ISSUE TITLE
Iteration over Set object with ES2016 is broken, because Set doesn't have length property

### ISSUE THREAD

**Issue by andrewpmontgomery** [angular/angular-cli#16229]
# 🐞 bug report

### Affected Package
Somewhere in the build process, after Typescript (tsc) but before AOT.

### Is this a regression?
Don't know.

### Description

Code which iterates over a `Set` is compiled as if Set has a `.length` property, but it doesn't, the correct property is `.size`. This silently breaks the logic.

This only happens when using `ng build`; it doesn't happen with `ng serve`.

## 🔬 Minimal Reproduction

Use `ng new` to create a minimal project.

Add the following to `app.component.ts`:
```typescript
export class AppComponent {
  sum = 0;
  constructor() {
    const numSet = new Set([12, 13, 17]);
    for (const num of numSet) {
      this.sum += num;
    }
    console.log(this.sum);
  }
}
```

In `tsconfig.json`, make just one change:
```
"target": "es2016",
```

Run `ng build` (not in prod mode).

Here's the compiled `main-es2016.js`:
```javascript
var AppComponent = /** @class */ (function () {
  function AppComponent() {
    this.sum = 0;
    var numSet = new Set([12, 13, 17]);
    for (var _i = 0, numSet_1 = numSet; _i < numSet_1.length; _i++) {
      var num = numSet_1[_i];
      this.sum += num;
    }
    console.log(this.sum);
  }
  return AppComponent;
}());
```

## 🔥 Exception or Error
`this.sum` should evaluate to 42; instead it evaluates to zero.
This happens because `numSet_1.length` is undefined. For the Set object, the compiler should have used the `.size` property, not `.length`.

## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 8.3.19
Node: 10.15.3
OS: win32 x64
Angular: 8.2.14
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.803.19
@angular-devkit/build-angular     0.803.19
@angular-devkit/build-optimizer   0.803.19
@angular-devkit/build-webpack     0.803.19
@angular-devkit/core              8.3.19
@angular-devkit/schematics        8.3.19
@angular/cli                      8.3.19
@ngtools/webpack                  8.3.19
@schematics/angular               8.3.19
@schematics/update                0.803.19
rxjs                              6.4.0
typescript                        3.5.3
webpack                           4.39.2
</code></pre>

**Anything else relevant?**
There are plenty of ways to work around this bug, but it's insidious so it really should be fixed.

I can't replicate this in Typescript (with `tsc` alone), so I assume the problem is in Angular, not in Typescript.

**Comment by andrewpmontgomery** [comment]
Looks like you're right, it's just falling back to ES5.
So it's actually a Typescript error in compiling to ES5, which I can replicate using `tsc` alone.
I'll raise this as a bug with the Typescript team.

**Comment by gkalpak** [comment]
(BTW, the TS behavior is intended. If you need to support iteration in <=ES5, you need to use [downlevelIteration][1].)

[1]: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-3.html#new---downleveliteration

**Comment by andrewpmontgomery** [comment]
I still can't replicate it with `tsc` alone. My guess is that it's a bug specific to the ES5/ES2015 differential bundle generation.

In file `main-es2015.js`, I get the correct output (almost the same as Typescript).
In file `main-es5.js`, it tries to use the `.length` property of the Set object.

In the compiled for...of loop, the undefined .length is effectively treated as zero. This isn't reported as an error in the browser, so it's particularly hard to discover.

I think this needs to be fixed in Angular.

**Comment by andrewpmontgomery** [comment]
@clydin - Thanks, that seems to do the trick.

**Comment by clydin** [comment]
Fixed via #16232
